samples/nrf9160/modem_shell/src/rest/rest_shell.c

```diff
@@ -0,0 +1,301 @@
+	"Simple REST shell client\n"
+	"Usage:\n"
+	"rest -d host [-m method] [-p port] [-s sec_tag] [-v verify_level] [-u url] [-t timeout_in_ms]\n"
+	"               [-H header1] [-H header2 ... header10] [-b body] [-l response_buffer_length]\n"
+	"               [-k] [-i sckt_id]\n"
+	"\n"
+	"  -d, --host,         Destination host/domain of the URL\n"
+	"Optionals:\n"
+	"  -m, --method,       HTTP method (one of the: \"get\" (default),\"head\",\n"
+	"                      \"post\",\"put\",\"delete\" or \"patch\")\n"
+	"  -p, --port,         Destination port (default: 80)\n"
+	"  -s, --sec_tag,      Used security tag for TLS connection\n"
+	"  -v, --peer_verify,  TLS peer verification level. None (0),\n"
+	"                      optional (1) or required (2). Default value is 2.\n"
+	"  -u, --url,          URL beyond host/domain (default: \"/index.html\")\n"
+	"  -t, --timeout,      Request timeout in seconds\n"
+	"                      (default: CONFIG_REST_CLIENT_REST_REQUEST_TIMEOUT)\n"
+	"  -H, --header,       Header including CRLF, for example:\n"
+	"                      -H \"Content-Type: application/json\\x0D\\x0A\"\n"
+	"  -b, --body,         Payload body, example: -b '{\"foo\":bar}'\n"
+	"  -l, --resp_len,     Length of the buffer reserved for the REST response\n"
+	"                      (default: 1024 bytes)\n"
+	"  -k, --keepalive,    Keep socket connection alive for upcoming requests\n"
+	"                      towards the ame server (default: false)\n"
+	"  -i, --sckt_id,      Use existing socket id (default: create a new connection)\n";
+
+/* Specifying the expected options (both long and short): */
+static struct option long_options[] = {
+	{ "host", required_argument, 0, 'd' },
+	{ "port", required_argument, 0, 'p' },
+	{ "url", required_argument, 0, 'u' },
+	{ "body", required_argument, 0, 'b' },
+	{ "resp_len", required_argument, 0, 'l' },
+	{ "header", required_argument, 0, 'H' },
+	{ "method", required_argument, 0, 'm' },
+	{ "sec_tag", required_argument, 0, 's' },
+	{ "peer_verify", required_argument, 0, 'v' },
+	{ "timeout", required_argument, 0, 't' },
+	{ "keepalive", no_argument, 0, 'k' },
+	{ "sckt_id", required_argument, 0, 'i' },
+	{ 0, 0, 0, 0 }
+};
+
+static void rest_shell_print_usage(const struct shell *shell)
+{
+	mosh_print_no_format(rest_shell_cmd_usage_str);
+}
+
+/*****************************************************************************/
+
+#define REST_REQUEST_MAX_HEADERS 10
+#define REST_RESPONSE_BUFF_SIZE 1024
+#define REST_DEFAULT_DESTINATION_PORT 80
+
+struct rest_header_list_item {
+	bool in_use;
+	char *header_str;
+};
+
+int rest_shell(const struct shell *shell, size_t argc, char **argv)
+{
+	int opt, i, len;
+	int ret = 0;
+	int long_index = 0;
+	bool show_usage = false;
+	struct rest_client_req_context req_ctx = { 0 };
+	struct rest_client_resp_context resp_ctx = { 0 };
+	/* Headers to request + 1 for NULL: */
+	char *req_headers[REST_REQUEST_MAX_HEADERS + 1] = { NULL };
```

Comment:
Do we really need +1 in the size? The reason I ask this is that the use of `req_headers` seems to always scroll through the table with condition `i < REST_REQUEST_MAX_HEADERS` and I couldn't find where we actually access the element `req_headers[REST_REQUEST_MAX_HEADERS]`.

Reply:
It is needed as the comment above says for ending NULL (it goes all the way to http_client) and we are initializing it here as NULL and intentionally don't touch it elsewhere.

Reply:
OK, so http_client lib needs the final NULL element. Please mention http_lib in the comment so it's very clear and nobody removes it in the future.